vim-patch:8.0.0074

DarkDeepBlue opened the pull request:
Problem:    Cannot make Vim fail on an internal error.
Solution:   Add IEMSG() and IEMSG2(). (Domenique Pelle)  Avoid reporting an
            internal error without mentioning where.

https://github.com/vim/vim/commit/95f096030ed1a8afea028f2ea295d6f6a70f466f

There is one special thing to consider while applying the patch. It makes it necessary to call `emsgf()` with an already prepared `va_list` from `iemsgf()` to avoid code duplication. Therefore I added `emsgfv()`, which takes an already prepared `va_list` and which is used in both, `emsgf()` and `iemsgf()` after preparing the `va_list`.

Any objections and comments?

DarkDeepBlue committed:
vim-patch:8.0.0074

Problem:    Cannot make Vim fail on an internal error.
Solution:   Add IEMSG() and IEMSG2(). (Domenique Pelle)  Avoid reporting an
            internal error without mentioning where.

https://github.com/vim/vim/commit/95f096030ed1a8afea028f2ea295d6f6a70f466f

Signed-off-by: Michael Schupikov <michael@schupikov.de>

KillTheMule commented:
You're missing some lints, you can do it locally via `make clint` or `make lint`.

DarkDeepBlue commented:
Thanks. I have already seen that, but some lints seemed not related to my code or at least I had the impression, so I corrected some of them but left the ones that are too far away from places I modified. After your comment I will address all of the lint issues now.

KillTheMule commented:
Yeah that's a common thing, the linter is pretty trigger happy since it looks at adjacent lines to your changes, too. Might cascade a bit, too :)

DarkDeepBlue commented:
I have addressed the lint errors, but how do I correct code like the following in `regexp.c`?
```
#define EMSG2_RET_FAIL(m, c) \
                       return (EMSG2((m), (c) ? "" : "\\"), \
                               rc_did_emsg = true, FAIL)
```
The according message is
```
Inner expression should be aligned as opening brace + 1 (+ 2 in case of {)  [whitespace/alignment] [2]
```
for the second line in the above macro, but the `return` is already aligned with the opening parentheses. Actually, I tried out every identation from 0 to beyond the last character on the previous line and always got the warning.

Would someone show me how to write such macros properly?

KillTheMule commented:
The linter surely isn't perfect, and heuristically I'd assume macros are a problem. You can disable linting selectively by using `// NOLINT(whitespace/alignment)` on the offending line. Not sure how to put comments into macros, though :)

justinmk commented:
If the linter violation is intentional, and there are many cases, usually it's preferred to just ignore the linter rather than adding very many "NOLINT" comments. Once this reaches master the linter will ignore those cases.

justinmk commented:
Do we need to add something in CMakeLists.txt like https://github.com/vim/vim/commit/95f096030ed1a8afea028f2ea295d6f6a70f466f#diff-f57f2991a6aa25fe45d8036c51bf8b4dR686 ?

justinmk commented on src/nvim/option.c:
```diff
@@ -2302,7 +2302,7 @@ set_string_option_direct (
     if (idx < 0) {  // Not found (should not happen).
-      EMSG2(_(e_intern2), "set_string_option_direct()");
+      internal_error("set_string_option_direct()");
       EMSG2(_("For option %s"), name);
```
this was also changed to IEMSG2 in the vim patch

justinmk commented:
missing:

- move e_intern2 up in globals.h
- change in if_cscope.c
- some changes in memfile.c, e.g. for `EMSGN(_("E323: line count wrong in block %" PRId64), bnum`

justinmk commented on src/nvim/message.c:
```diff
@@ -583,19 +583,82 @@ void emsg_invreg(int name)
+/// Give an "Internal error" message.
+void internal_error(char *where)
+{
+    IEMSG2(_(e_intern2), where);
```
wrong indent level. same for the others

justinmk commented on src/nvim/message.c:
```diff
@@ -583,19 +583,82 @@ void emsg_invreg(int name)
+/// Same as emsgn(...) but abort on error when ABORT_ON_INTERNAL_ERROR is
+/// defined. It is used for internal errors only, so that they can be
+/// detected when fuzzing vim.
+void iemsgn(const char *s, long n)
```
this isn't needed, right?

justinmk commented on src/nvim/message.c:
```diff
@@ -583,19 +583,82 @@ void emsg_invreg(int name)
 
+/// Print an error message with one "%ld" and one long int argument.
+/// This is not in message.c to avoid a warning for prototypes.
+int emsgn(const char *s, long n)
```
this isn't used anywhere

jamessan committed:
lint
